CONTRIBUTING.md

```diff
@@ -1,25 +1,58 @@
+4.	Java version.
+It is also helpful to mention the JVM parameters.
+5.	Operating system.
+If it is Linux, kernel version is helpful.
 6.	Logs and stack traces, if available.
 7.	Detailed description of the steps to reproduce your issue.
-8.	Unit test with the `hazelcast.xml` file. If you could include a unit test which reproduces your issue, we would be grateful.
-9.	If available, integration module versions (e.g. Tomcat, Jetty, Spring, Hibernate). Also, include their  detailed configuration information such as `web.xml`, Hibernate configuration and `context.xml` for Spring.
-
-## Pull Requests
-Thanks for creating your pull request (PR).
-1.	Contributions are submitted, reviewed, and accepted using the pull requests on GitHub.
-2.	Our bot will automatically ask you on the PR to sign the Hazelcast Contributor Agreement.
-3.	Try to make clean commits that are easily readable (including descriptive commit messages).
-4.	The latest changes are in the **master** branch.
-5.	Before you push, run the command `mvn clean package -P checkstyle,spotbugs` in your terminal and fix the CheckStyle and SpotBugs errors (if any). Push your PR once it is free of CheckStyle and SpotBugs errors.
-6.	Please keep your PRs as small as possible, i.e. if you plan to perform a huge change, do not submit a single and large PR for it. For an enhancement or larger feature, you can create a GitHub issue first to discuss.
+8.	Unit test with the `hazelcast.xml` file.
+If you could include a unit test which reproduces your issue, it would speed the resolution time.
+9.	If available, integration module versions (_e.g._ Tomcat, Jetty, Spring, Hibernate).
+Also, include their  detailed configuration information such as `web.xml`, Hibernate configuration and `context.xml` for Spring.
+
+## Pull requests
+
+Thanks a lot for creating your <abbr title="Pull Request">PR</abbr>!
+
+A PR can target many different subjects:
+
+* [Documentation](https://github.com/hazelcast/hazelcast-reference-manual):
+either fix typos or improve the documentation as a whole
+* Fix a bug
+* Add a feature
+* Add additional tests to improve the test coverage, or fix flaky tests
+* Anything else that makes Hazelcast better!
+
+All PRs follow the same process:
+
+1.	Contributions are submitted, reviewed, and accepted using the PR system on GitHub.
+2.	For first time contributors, our bot will automatically ask you to sign the Hazelcast Contributor Agreement on the PR.
+3.	The latest changes are in the `master` branch.
+4.	Make sure to design clean commits that are easily readable.
+That includes descriptive commit messages.
+5.	Please keep your PRs as small as possible, _i.e._ if you plan to perform a huge change, do not submit a single and large PR for it.
+For an enhancement or larger feature, you can create a GitHub issue first to discuss.
+6.	Before you push, run the command `mvn clean package -P checkstyle,spotbugs` in your terminal and fix the CheckStyle and SpotBugs errors if any.
```

Comment:
We should suggest skipping tests as they take hours.

```suggestion
6.	Before you push, run the command `mvn clean package -Pcheckstyle,spotbugs -DskipTests` in your terminal and fix the CheckStyle and SpotBugs errors if any.
```

Reply:
You don't want people to fail their tests before pushing? 🤔

Reply:
Usually they can run tests they've created and possibly `QuickTest` tests in the same package. For instance, when adding a feature on top of `IQueue`, running all queue tests. But often times it takes tens of minutes or even hours to run tests on a home computer as running them on our lab machines takes up to 40 minutes.

Reply:
That's kind of my point. The default set of tests executed by `mvn clean package` will take ~5 hours. Noone will ever run that and definitely not repeatedly. We have PR builder, we can verify that on our side.

Reply:
Ok.

1. As a former engineer (and build master), I don't understand why we haven't a clear separation between "quick" tests in the `test` phase, and other tests in the `integration-test` phase. I understand once those are here, it's time-consuming to triage, but the benefits are huge.
2. Why do we ask for `package` instead of just `compile`? Do we do some black magic in `package`? (asking for a friend, of course)

Reply:
1. We actually have the categories, quick and slow tests. And those quick tests are actually run in `test` phase. By a quick test we mean that the test itself is running pretty quickly. But we have so many of them, so it takes hours. About your triage, how would you then select the tests that should be run by default? Hazelcast is a huuuuge project, how do you pick the testsuite that actually gives you some sort of confidence? This is what the default `test` does, but then, it takes hours.
2. I guess the intention is that the `package` will actually run the tests :)

Having this discussion, I realize that maybe my initial proposal is wrong. We shouldn't advertise to skip the tests right away. We kind of know that nobody will ever run it, but let's not advertise and let them skip it themselves. In addition to actually have some useful information, we could add a note how to run only some specific tests, i. e. using `-Dtest=...`. WDYT?

Reply:
I think that between not running tests at all and running all tests, there must be a sweet spot. That's actually a good idea to provide a basic test harness.

Reply:
I do agree it would be nice (or simply interesting?) if there was a category of "smoke" tests that the user might run locally. Thing is, since I've joined, there were already more than 20k tests in the `QuickTest` category so it would first take an effort to identify the basic ones which are worth their own category to run locally. 

Secondly, the PR builder always took about 30-40 minutes so most of the time, it was simpler for everyone involved to simply push the commit and get feedback from all 20k - 30k tests. Running less test (e.g. 1k) might feel good (as if nothing has been broken) but I'm not sure what kind of coverage we would have with that set of tests, and again the full suite would be run as soon as the commit was pushed, which would uncover many more issues.

Also, a lot of the times community PRs fail on checkstyle and findbugs. We say users should run them, but they don't. So really, I don't expect users to run even 1k tests locally :)

Finally, I do see some room for improvement. I don't believe we actually need about 30k tests that we have now and I'm sure there is lots of duplication. What would be great if we could clean up the test codebase to decrease the number of tests, decrease the PR builder time and maybe even make it possible to run some sensible set of tests locally. But, it's a lot of work and we'll probably iteratively clean up tests as we work on the relevant code and I don't think we'll reach the final state any time soon.